nx/exercises/exercises-1-20.livemd

```diff
@@ -0,0 +1,490 @@
+
+  </div>
+</details>
+
+<!-- livebook:{"break_markdown":true} -->
+
+**16. Add a border of 0 around `tensor_16` (end result will be a 5x5 tensor). (★☆☆)**
+
+```elixir
+tensor_16 = Nx.broadcast(1, {3, 3})
+# Add your solution here.
+:ok
+```
+
+<details class="rounded-lg border-2 border-green-bright-300 p-2">
+  <summary class="cursor-pointer font-bold mx-2">Example solution</summary>
+  <div class="p-2">
+
+  ```elixir
+  Nx.pad(tensor_16, 0, [{1, 1, 0}, {1, 1, 0}])
+  ```
+
+  </div>
+</details>
+
+<!-- livebook:{"break_markdown":true} -->
+
+**17. Determine the results of the following expressions. (★☆☆)**
+
+```elixir
+nan = Nx.Constants.nan()
+Nx.multiply(0, nan)
+Nx.equal(nan, nan)
+Nx.greater(nan, nan)
+Nx.subtract(nan, nan)
+
+# Add your solution here.
+:ok
+```
+
+<details class="rounded-lg border-2 border-green-bright-300 p-2">
+  <summary class="cursor-pointer font-bold mx-2">Example solution</summary>
+  <div class="p-2">
+
+  ```
+  #Nx.Tensor<
+    f32
+    NaN
+  >
+  #Nx.Tensor<
+    u8
+    0
+  >
+  #Nx.Tensor<
+    u8
+    0
+  >
+  #Nx.Tensor<
+    f32
+    NaN
+  >
+  ```
+
+  </div>
+</details>
+
+<!-- livebook:{"break_markdown":true} -->
+
+**18. Create a 5x5 tensor with values 1,2,3,4 just below the diagonal. (★☆☆)**
+
+```elixir
+# Add your solution here.
+:ok
+```
+
+<details class="rounded-lg border-2 border-green-bright-300 p-2">
+  <summary class="cursor-pointer font-bold mx-2">Example solution</summary>
+  <div class="p-2">
+
+  ```elixir
+  Nx.tensor([1, 2, 3, 4])
+  |> Nx.make_diagonal(offset: -1)
+  ```
+
+  </div>
+</details>
+
+<!-- livebook:{"break_markdown":true} -->
+
+**19. Create a 8x8 tensor of 0 and 1 in a checkerboard pattern with 0 as the first element. (★☆☆)**
+
+```elixir
+# Add your solution here.
+:ok
+```
+
+<details class="rounded-lg border-2 border-green-bright-300 p-2">
+  <summary class="cursor-pointer font-bold mx-2">Example solution</summary>
+  <div class="p-2">
+
+  ```elixir
+  Nx.tensor([[0, 1], [1, 0]])
+  |> Nx.tile([4, 4])
+  ```
+
+  </div>
+</details>
+
+<!-- livebook:{"break_markdown":true} -->
+
+**20. Find the (x,y,z) index ([row major](https://en.wikipedia.org/wiki/Row-_and_column-major_order)) of the 100th element of a 6x7x8 tensor. (★★☆)**
+
+```elixir
+# Add your solution here.
+:ok
+```
+
+<details class="rounded-lg border-2 border-green-bright-300 p-2">
+  <summary class="cursor-pointer font-bold mx-2">Example solution</summary>
+  <div class="p-2">
+
+  ```elixir
+  defmodule Exercise20 do
+    import Nx.Defn
+
+    deftransform solve(shape, target) do
+      dims = Tuple.to_list(shape)
+      
+      dims
+      |> Enum.map_reduce({target, Enum.product(dims)}, fn dim, {remaining, step_size} ->
+        step_size = div(step_size, dim)
+        num_steps = div(remaining, step_size)
+        {num_steps, {remaining - num_steps * step_size, step_size}}
+      end)
+      |> then(fn {indices, {0, 1}} -> indices end)
+    end
+  end
+
+  Exercise20.solve({6, 7, 8}, 100)
+  ```
```

Comment:
````suggestion
  ```

  > **Note:** functions defined with `deftransform` can be invoked inside `defn` and they are oftentimes used to compute tensor shapes. Inside `deftransform` you can call any other modules, such as `Enum` and `Tuple`, which is not the possible inside `defn`.
````

Reply:
I think I want to replace this exercise with something simpler. These early exercises are mostly about getting the reader to lookup the relevant function. In Numpy, you simply call `unravel`. Adhering to the original isn't as important as keeping the difficulty consistent IMO.

WDYT? \cc @polvalente

Reply:
Exercise 21 in the kaggle notebook is:

> 21. Create a checkerboard 8x8 matrix using the tile function (★☆☆)

Since that was actually my solution to 19, how about for 20:

<blockquote>

20. Produce the same checkerboard pattern from exercise 19, but use [Nx.window_reduce](https://hexdocs.pm/nx/Nx.html#window_reduce/5) instead. (★☆☆)

<details>
<summary>Example solution</summary>

```elixir
Nx.window_reduce(Nx.iota({9, 9}), 0, {2, 2}, [], fn x, _ -> Nx.remainder(x, 2) end)
```

</details>
</blockquote>

(This may be a 2-star since `window_reduce` is a little complicated and it's not obvious how to use it here.)

Reply:
Just use remainder on the iota as I had initially suggested. We shouldn't incentivize the usage of window_reduce because it's far less efficient than alternatives when available

Reply:
I mentioned before that doesn't work because the checkerboard is even by even:

```elixir
Nx.iota({8, 8}) |> Nx.remainder(2)
#Nx.Tensor<
  s64[8][8]
  [
    [0, 1, 0, 1, 0, 1, 0, 1],
    [0, 1, 0, 1, 0, 1, 0, 1],
    [0, 1, 0, 1, 0, 1, 0, 1],
    [0, 1, 0, 1, 0, 1, 0, 1],
    [0, 1, 0, 1, 0, 1, 0, 1],
    [0, 1, 0, 1, 0, 1, 0, 1],
    [0, 1, 0, 1, 0, 1, 0, 1],
    [0, 1, 0, 1, 0, 1, 0, 1]
  ]
>
```

You could do a 9x9 and slice I suppose. Is that what you mean?

Reply:
Ah, makes sense. You could do that slicing , or go with a more mathematical approach that relies on broadcasting. Although maybe you could just use tile and skip the other exercise. PS: I haven't checked the solution below, take it with a grain of salt. 

t = Nx.iota({8, 1}) |> Nx.remainder(2)

Nx.transpose(t) |> Nx.add(t) |> Nx.remainder(2)

Reply:
Your solution works! :)

So here's what I'm thinking now: we just say "Make a checkerboard _without_ `Nx.tile`." Should get the creative juices flowing and force you to read the docs for ideas.

Reply:
How about this: suggest Nx.tile here, and the version without Nx.tile later on?

Reply:
Hm I think we're having a two slightly different conversations.

I'm talking about replacing exercise 20 since Nx doesn't have an `unravel` equivalent. I was suggesting we replace 20 with a twist on 19 since 19 is the one where we ask for a checkerboard with `Nx.tile`.

Sorry if I'm misunderstanding! It may be too late to be on Github.

Reply:
Yeah, to be clear:

I think the first checkerboard exercise should be with tile, and then the next should be "ok, now do it without tile", maybe hinting on broadcasting or something like that!

Reply:
[58eb8b1](https://github.com/elixir-nx/nx/pull/1363/commits/58eb8b1e431653316828a863ddfcdc978e31d5d1)